doc/statemachines.tex

```diff
@@ -0,0 +1,310 @@
+  which also means that client and server simultaneously transit to new states.
+  The global state is not observable by the client nor the server.
+  Possible interpretations: 
+  A sequence of states is a set of points in time that a correct run of the protocol passes through.
+  (todo: draw diagram)
+\item[Messages/Events]
+  A tuple $(lable,data)$. Can be serialized transferred over the network and de-serialized.
+  (authenticity of messages not handled here)
+\item[Transitions]
+  The document describes the global/abstract state transitions of the protocol
+  in terms of messages and data that the client and the server exchange.
+  In sync with state transitions the  client and server process the data locally to compute the
+  the actual "outcome" of a protocol-run. (For example "update the block chain").
+  Parts of the local computations are specific to the implementation language and the underlying
+  data structures, but they are also relevant for the protocol (to determine the next transition).
+\item[Initial state]
+\item[Final state \& termination]
+\end{description}
+  
+\subsubsection{Protocol Description}
+
+
+\section{The Chain Synchronization Protocol}
+This section describes the Chain Synchronization Protocol.
+
+\begin{description}
+\item[purpose of the protocol]
+The Chain Synchronization Protocol is used by the block chain consumer
+to synchronize its block chain with the block chain producer.
+\item[Who is communicating]
+  A node communicates with several upstream and downstream node.
+  The node runs an independent agents for every other node
+  it communicates with. The typically a node runs agents where it acts as a server and also instances
+  where it acts as a client.
+\item[Where is located in the protocol stack ?]
+\item[What are the interfaces to the lower layers of the stack ?]
+  The chain synchronization protocol can send and receive messages. (via a point to point channel.)
+\item[Initialization of an instance of the protocol]
+\end{description}
+
+\subsubsection{State machine}
+
+\begin{tabular}{|l|l|}
+  \hline
+  \multicolumn{2}{|c|}{States} \\ \hline
+  Name  & Agency \\ \hline \hline
+  Idle       & client \\ \hline
+  CanAwait   & server \\ \hline
+  MustReply  & server \\ \hline  
+  Intersect  & server \\ \hline
+  Done       &        \\ \hline
+  \hline
+\end{tabular}
+
+\begin{tabular}{|l|l|l|l|}
+  \hline
+  \multicolumn{4}{|c|}{Transitions} \\ \hline
+  message/event      & parameter     & from        & to       \\ \hline\hline
+  RequestNext        &               & Idle        & CanAwait \\ \hline
+  AwaitReply         &               & CanAwait    & MustReply \\ \hline
+  RollForward        & $h$,$p$       & CanAwait    & Idle \\ \hline
+  RollForward        & $h$,$p$       & MustReply   & Idle \\ \hline
+  RollBackward       & $p$,$p$       & CanAwait    & Idle \\ \hline
+  RollBackward       & $p$,$p$       & MustReply   & Idle \\ \hline  
```

Comment:
If you'll check the codec, you will see that both `RollBackward` messages will have the same encoding (so there is actually a single one), maybe it's worth to preserve that `RollBackward` is polymorphic? What do you think?

Reply:
There really are 2 different roll-backward transitions: they each have different initial states, so they must be different.

Different transitions may agree on _encoding_ but still be different. The encoding need only be injective _for a given initial state_.